.cloudbuild/scripts/cmd/integration-tests/main.go

```diff
@@ -0,0 +1,189 @@
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"syscall"
+
+	"github.com/gravitational/teleport/.cloudbuild/scripts/internal/changes"
+	"github.com/gravitational/teleport/.cloudbuild/scripts/internal/etcd"
+	"github.com/gravitational/trace"
+)
+
+const (
+	gomodcacheDir = ".gomodcache-ci"
+	nonrootUID    = 1000
+	nonrootGID    = 1000
+)
+
+// main is just a stub that prints out an error message and sets a nonzero exit
+// code on failure. All of the work happens in `innerMain()`.
+func main() {
+	if err := innerMain(); err != nil {
+		log.Fatalf("FAILED: %s", err.Error())
+	}
+}
+
+type commandlineArgs struct {
+	workspace    string
+	targetBranch string
+	commitSHA    string
+	skipChown    bool
+}
+
+func parseCommandLine() (commandlineArgs, error) {
+	args := commandlineArgs{}
+
+	flag.StringVar(&args.workspace, "w", "", "Fully-qualified path to the build workspace")
+	flag.StringVar(&args.targetBranch, "t", "", "The PR's target branch")
+	flag.StringVar(&args.commitSHA, "c", "", "The PR's latest commit SHA")
+	flag.BoolVar(&args.skipChown, "skip-chown", false, "Skip reconfiguring the workspace for a nonroot user.")
+
+	flag.Parse()
+
+	if args.workspace == "" {
+		return args, trace.Errorf("workspace path must be set")
+	}
+
+	var err error
+	args.workspace, err = filepath.Abs(args.workspace)
+	if err != nil {
+		return args, trace.Wrap(err, "Unable to resole absolute path to workspace")
+	}
+
+	if args.targetBranch == "" {
+		return args, trace.Errorf("target branch must be set")
+	}
+
+	if args.commitSHA == "" {
+		return args, trace.Errorf("commit must be set")
+	}
+
+	return args, nil
+}
+
+// innerMain parses the command line, performs the highlevel docs change check
+// and creates the marker file if necessary
+func innerMain() error {
+	args, err := parseCommandLine()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	gomodcache := fmt.Sprintf("GOMODCACHE=%s", path.Join(args.workspace, gomodcacheDir))
+
+	log.Println("Analysing code changes")
+	ch, err := changes.Analyze(args.workspace, args.targetBranch, args.commitSHA)
+	if err != nil {
+		return trace.Wrap(err, "Failed analyzing code")
+	}
+
+	hasOnlyDocChanges := ch.Docs && (!ch.Code)
+	if hasOnlyDocChanges {
+		log.Println("No non-docs changes detected. Skipping tests.")
+		return nil
+	}
+
+	log.Printf("Running root-only integration tests...")
+	err = runRootIntegrationTests(args.workspace, gomodcache)
+	if err != nil {
+		return trace.Wrap(err, "Root-only integration tests failed")
+	}
+
+	if !args.skipChown {
+		// We run some build steps as root and others as a non user, and we
+		// want the nonroot user to be able to manipulate the artifacts
+		// created by root, so we `chown -R` the whole workspace to allow it.
+		log.Printf("Reconfiguring workspace for nonroot user")
+		err = chownR(args.workspace, nonrootUID, nonrootGID)
+		if err != nil {
+			return trace.Wrap(err, "failed reconfiguring workspace")
+		}
+	}
+
+	log.Printf("Starting etcd...")
```

Comment:
Are the `ectd` tests the only tests we run as non-root?

Reply:
We run a heap of tests as nonroot - the `etcd` tests just happen to be among them. I tried running `etcd` as root and the tests as nonroot. Unfortunately this produces a heap of spurious error messages[1] when running the tests that I did not want confusing people in the logs. 

1. The files created by `etcd` live inside the directory searched by `go list ./...` and so on to generate the list of package names, and so making them owned by `root` produces a heap of diagnostic warnings that would pollute the build log and just confuse people when they are trying to work out why their build failed. 

...Perhaps this should be a comment in the program :-)